File: java/formula-java/src/main/java/org/formula/parse/shuntingyard/ShuntingYardParser.java
package org.formula.parse.shuntingyard;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Stack;
import java.util.function.Supplier;
import org.formula.Resolvable;
import org.formula.ResolvedValue;
import org.formula.parse.Parser;
import org.formula.parse.tree.NodeExpression;
import org.formula.parse.tree.TokenTree;
import org.formula.util.Lambda2;

public class ShuntingYardParser implements Parser {
    private static final String PARAMETERS_START_SYMBOL = "(";
    private static final String PARAMETERS_END_SYMBOL = ")";
    private static final String PARAMETERS_SEPARATOR_SYMBOL = ";";
    private static final Literal PARAMETERS_START_TOKEN = Literal.of(PARAMETERS_START_SYMBOL);
    private static final Literal PARAMETERS_END_TOKEN = Literal.of(PARAMETERS_END_SYMBOL);
    private static final Literal PARAMETERS_SEPARATOR_TOKEN = Literal.of(PARAMETERS_SEPARATOR_SYMBOL);

    private final TokenTree<Node> tokenTree;

    public static ShuntingYardParser create() {
        return new ShuntingYardParser();
    }

    public ShuntingYardParser() {
        tokenTree = TokenTree.<Node>create()
                .ignoreWhitespaces()
                .add(NodeExpression.INTEGER, token -> Term.of(Integer.parseInt(token)))
                .add(NodeExpression.DECIMAL, token -> Term.of(Double.parseDouble(token)))
                .add(PARAMETERS_START_SYMBOL, ignored -> PARAMETERS_START_TOKEN)
                .add(PARAMETERS_END_SYMBOL, ignored -> PARAMETERS_END_TOKEN)
                .add(PARAMETERS_SEPARATOR_SYMBOL, ignored -> PARAMETERS_SEPARATOR_TOKEN)
                .add(NodeExpression.literal("\"", "\"", "\\\""),
                        quote -> Term.of(quote.substring(1, quote.length() - 1), "\"", "\""))
                .add(NodeExpression.literal("'", "'", "\\'"),
                        quote -> Term.of(quote.substring(1, quote.length() - 1), "'", "'"));
    }

    public ShuntingYardParser biOperator(String symbol, Operator1 unaryOperator, Operator2 binaryOperator) {
        tokenTree.add(symbol, token -> new BiOperatorFunction(symbol, unaryOperator, binaryOperator));
        return this;
    }

    public ShuntingYardParser operator(String symbol, int precedence, Associativity associativity,
            OperatorFunction1 fn) {
        tokenTree.add(symbol, token -> new Operator1(symbol, precedence, associativity, fn));
        return this;
    }

    public ShuntingYardParser operator(String symbol, int precedence, Associativity associativity,
            OperatorFunction2 fn) {
        tokenTree.add(symbol, token -> new Operator2(symbol, precedence, associativity, fn));
        return this;
    }

    public ShuntingYardParser function(String name, OperatorFunction0 fn) {
        tokenTree.add(name, token -> new Function0(name, fn));
        return this;
    }

    public ShuntingYardParser function(String name, OperatorFunction1 fn) {
        tokenTree.add(name, token -> new Function1(name, fn));
        return this;
    }

    public ShuntingYardParser function(String name, OperatorFunction2 fn) {
        tokenTree.add(name, token -> new Function2(name, fn));
        return this;
    }

    public ShuntingYardParser function(String name, OperatorFunction3 fn) {
        tokenTree.add(name, token -> new Function3(name, fn));
        return this;
    }

    public ShuntingYardParser function(String name, OperatorFunctionN fn) {
        tokenTree.add(name, token -> new FunctionN(name, fn));
        return this;
    }

    public ShuntingYardParser variable(String identifier, VariableResolver variableResolver) {
        NodeExpression variableExpression = NodeExpression.of(
                NodeExpression.term(identifier),
                NodeExpression.ALPHA,
                NodeExpression.optional(NodeExpression.KEY)
        );
        tokenTree.add(variableExpression,
                key -> new Variable(key, (context, k) -> variableResolver.resolve(context, k.substring(identifier.length()))));
        return this;
    }

    public ShuntingYardParser variable(String prefix, String suffix, VariableResolver variableResolver) {
        NodeExpression variableExpression = NodeExpression.of(
                NodeExpression.term(prefix),
                NodeExpression.ALPHA,
                NodeExpression.optional(NodeExpression.KEY),
                NodeExpression.term(suffix)
        );
        tokenTree.add(variableExpression,
                key -> new Variable(key, (context, k) -> variableResolver.resolve(context,
                        k.substring(prefix.length(), k.length() - suffix.length()))));
        return this;
    }

    public ShuntingYardParser comment(String prefix, String suffix, Lambda2<ResolvedValue, String, ResolvedValue> fn) {
        NodeExpression commentExpression = NodeExpression.literal(prefix, suffix);
        tokenTree.add(commentExpression, token -> Comment.of(token, fn));
        return this;
    }

    public ShuntingYardParser term(String text, Supplier<ResolvedValue> extractor) {
        NodeExpression termExpression = NodeExpression.term(text);
        tokenTree.add(termExpression, key -> Term.of(extractor.get()));
        return this;
    }

    @Override
    public Resolvable parse(String text) {
        Stack<Node> operatorStack = new Stack<>();
        Stack<Node> outputBuffer = new Stack<>();
        Stack<Integer> arityStack = new Stack<>();

        List<Node> tokens = tokenTree.parse(text);

        for (int i = 0; i < tokens.size(); i++) {
            Node token = tokens.get(i);
            Node previous = i > 0 ? tokens.get(i-1) : null;

            if (token instanceof BiOperatorFunction func) {
                token = func.binaryOperator();
                if (previous == null || previous instanceof Operator
                        || Objects.equals(previous, PARAMETERS_START_TOKEN)
                        || Objects.equals(previous, PARAMETERS_SEPARATOR_TOKEN)) {
                    token = func.unaryOperator();
                }
            }

            if (token instanceof Operator operator) {
                if (!operatorStack.isEmpty()) {
                    Node top = operatorStack.peek();
                    if (top instanceof Operator topOperator) {
                        if ((operator.precedence() < topOperator.precedence())
                                || (operator.associativity() == Associativity.LEFT
                                && operator.precedence() == topOperator.precedence())) {
                            operatorStack.pop();
                            outputBuffer.push(top);
                        }
                    }
                }

                operatorStack.push(operator);
                continue;
            }

            if (token instanceof Function) {
                operatorStack.push(token);
                arityStack.push(1);
                continue;
            }

            if (token instanceof Variable) {
                outputBuffer.push(token);
                continue;
            }

            if (token instanceof Term) {
                outputBuffer.push(token);
                continue;
            }

            if (token instanceof Comment) {
                outputBuffer.push(token);
                continue;
            }

            if (token instanceof Literal literal) {
                switch (literal.value()) {
                    case " ":
                    case "{":
                    case "}":
                        // ignore
                        break;
                    case PARAMETERS_SEPARATOR_SYMBOL:
                        arityStack.push(arityStack.pop() + 1);
                        while (!operatorStack.isEmpty()) {
                            Node next = operatorStack.pop();
                            if (next.equals(PARAMETERS_START_TOKEN)) {
                                operatorStack.push(next);
                                break;
                            }
                            outputBuffer.push(next);
                        }
                        break;
                    case PARAMETERS_START_SYMBOL:
                        operatorStack.push(token);
                        break;
                    case PARAMETERS_END_SYMBOL:
                        while (!operatorStack.isEmpty()) {
                            var next = operatorStack.pop();
                            if (next.equals(PARAMETERS_START_TOKEN)) {
                                break;
                            }
                            outputBuffer.push(next);
                        }

                        if (!operatorStack.isEmpty() && operatorStack.peek() instanceof Function) {
                            if (operatorStack.peek() instanceof FunctionN) {
                                int arity = arityStack.pop();
                                arity = previous.equals(PARAMETERS_START_TOKEN) ? 0 : arity;
                                outputBuffer.push(new Arity(arity));
                            }
                            outputBuffer.push(operatorStack.pop());
                        }
                        break;
                    default:
                        outputBuffer.push(token);
                }
            }
        }

        while (!operatorStack.isEmpty()) {
            outputBuffer.push(operatorStack.pop());
        }
        return new ShuntingYard(Arrays.asList(outputBuffer.toArray()), text);
    }
}

File: cs/formula-cs/Formula/ShuntingYard/ShuntingYardParser.cs
using Formula.TokenTree;

namespace Formula.ShuntingYard;

public class ShuntingYardParser
{
    private const string ParametersStartSymbol = "(";
    private const string ParametersEndSymbol = ")";
    private const string ParametersSeparatorSymbol = ";";
    private static readonly Literal ParametersStartToken = new Literal(ParametersStartSymbol);
    private static readonly Literal ParametersEndToken = new Literal(ParametersEndSymbol);
    private static readonly Literal ParametersSeparatorToken = new Literal(ParametersSeparatorSymbol);
    
    private readonly TokenTree<INode> _tokenTree;

    public static ShuntingYardParser Create()
    {
        return new ShuntingYardParser();
    }

    private ShuntingYardParser()
    {
        _tokenTree = TokenTree<INode>.Create()
            .IgnoreWhitespaces()
            .Add(NodeExpression.Integer, token => new Term(ResolvedValue.Of(int.Parse(token))))
            .Add(NodeExpression.Decimal, token => new Term(ResolvedValue.Of(double.Parse(token))))
            .Add(ParametersStartSymbol, _ => ParametersStartToken)
            .Add(ParametersEndSymbol, _ => ParametersEndToken)
            .Add(ParametersSeparatorSymbol, _ => ParametersSeparatorToken)
            .Add(NodeExpression.Literal("\"", "\"", "\\\""),
                quote => new Term(ResolvedValue.Of(quote.Substring(1, quote.Length - 2)), "\"", "\""))
            .Add(NodeExpression.Literal("'", "'", "\\'"),
                quote => new Term(ResolvedValue.Of(quote.Substring(1, quote.Length - 2)), "'", "'"));
    }

    public ShuntingYardParser Operator(string symbol, int precedence, Associativity associativity,
        Func<ResolvedValue> fn)
    {
        _tokenTree.Add(symbol, _ => new Operator0(symbol, precedence, associativity, fn));
        return this;
    }

    public ShuntingYardParser Operator(string symbol, int precedence, Associativity associativity,
        Func<ResolvedValue, ResolvedValue> fn)
    {
        _tokenTree.Add(symbol, _ => new Operator1(symbol, precedence, associativity, fn));
        return this;
    }

    public ShuntingYardParser Operator(string symbol, int precedence, Associativity associativity,
        Func<ResolvedValue, ResolvedValue, ResolvedValue> fn)
    {
        _tokenTree.Add(symbol, _ => new Operator2(symbol, precedence, associativity, fn));
        return this;
    }

    public ShuntingYardParser BiOperator(string symbol, Operator1 unaryOperator, Operator2 binaryOperator)
    {
        _tokenTree.Add(symbol, _ => new BiOperator(symbol, unaryOperator, binaryOperator));
        return this;
    }

    public ShuntingYardParser Function(string name, Func<ResolvedValue> fn)
    {
        _tokenTree.Add(name, _ => new Function0(name, fn));
        return this;
    }

    public ShuntingYardParser Function(string name, Func<ResolvedValue, ResolvedValue> fn)
    {
        _tokenTree.Add(name, _ => new Function1(name, fn));
        return this;
    }

    public ShuntingYardParser Function(string name, Func<ResolvedValue, ResolvedValue, ResolvedValue> fn)
    {
        _tokenTree.Add(name, _ => new Function2(name, fn));
        return this;
    }

    public ShuntingYardParser Function(string name, Func<ResolvedValue, ResolvedValue, ResolvedValue, ResolvedValue> fn)
    {
        _tokenTree.Add(name, _ => new Function3(name, fn));
        return this;
    }

    public ShuntingYardParser Function(string name, Func<IEnumerable<ResolvedValue>, ResolvedValue> fn)
    {
        _tokenTree.Add(name, _ => new FunctionN(name, fn));
        return this;
    }

    public ShuntingYardParser Variable(string identifier, VariableResolver variableResolver)
    {
        var variableExpression = NodeExpression.Of(
            NodeExpression.Term(identifier),
            NodeExpression.Alpha,
            NodeExpression.Optional(NodeExpression.Key)
        );
        _tokenTree.Add(variableExpression,
            key => new Variable(key, (context, k) => variableResolver(context, k[identifier.Length..])));
        return this;
    }

    public ShuntingYardParser Variable(string prefix, string suffix, VariableResolver variableResolver)
    {
        var variableExpression = NodeExpression.Of(
            NodeExpression.Term(prefix),
            NodeExpression.Alpha,
            NodeExpression.Optional(NodeExpression.Key),
            NodeExpression.Term(suffix)
        );
        _tokenTree.Add(variableExpression,
            key => new Variable(key, (context, k) => variableResolver(context,
                k.Substring(prefix.Length, k.Length - prefix.Length - suffix.Length))));
        return this;
    }

    public ShuntingYardParser Comment(string prefix, string suffix, Func<ResolvedValue, string, ResolvedValue> fn)
    {
        var commentExpression = NodeExpression.Literal(prefix, suffix);
        _tokenTree.Add(commentExpression, token => new Comment(token, fn));
        return this;
    }

    public ShuntingYardParser Term(string text, Func<ResolvedValue> extractor)
    {
        var termExpression = NodeExpression.Term(text);
        _tokenTree.Add(termExpression, _ => new Term(extractor.Invoke()));
        return this;
    }

    public IResolvable Parse(string text)
    {
        var operatorStack = new Stack<INode>();
        var outputBuffer = new Stack<INode>();
        var arityStack = new Stack<int>();

        var tokens = _tokenTree.Parse(text);
        
        for (var i = 0; i < tokens.Count; i++)
        {
            var token = tokens[i];
            var previous = i > 0 ? tokens[i - 1] : null;

            if (token is BiOperator biOp)
            {
                token = biOp.BinaryOperator;
                if (previous is null or IOperator || previous.Equals(ParametersStartToken) || previous.Equals(ParametersSeparatorToken))
                {
                    token = biOp.UnaryOperator;
                }
            }
            
            if (token is IOperator op)
            {
                if (operatorStack.Count > 0)
                {
                    var top = operatorStack.Peek();
                    if (top is IOperator topOperator)
                    {
                        if (op.Precedence < topOperator.Precedence
                            || (op.Associativity == Associativity.Left
                                && op.Precedence == topOperator.Precedence))
                        {
                            operatorStack.Pop();
                            outputBuffer.Push(top);
                        }
                    }
                }

                operatorStack.Push(op);
                continue;
            }

            if (token is IFunction)
            {
                operatorStack.Push(token);
                arityStack.Push(1);
                continue;
            }

            if (token is Variable)
            {
                outputBuffer.Push(token);
                continue;
            }

            if (token is Term)
            {
                outputBuffer.Push(token);
                continue;
            }

            if (token is Comment)
            {
                outputBuffer.Push(token);
                continue;
            }

            if (token is Literal literal)
            {
                switch (literal.Value)
                {
                    case " ":
                    case "{":
                    case "}":
                        // ignore
                        break;
                    case ParametersSeparatorSymbol:
                        arityStack.Push(arityStack.Pop() + 1);
                        while (operatorStack.Count > 0)
                        {
                            var next = operatorStack.Pop();
                            if (next.Equals(ParametersStartToken))
                            {
                                operatorStack.Push(next);
                                break;
                            }

                            outputBuffer.Push(next);
                        }

                        break;
                    case ParametersStartSymbol:
                        operatorStack.Push(token);
                        break;
                    case ParametersEndSymbol:
                        while (operatorStack.Count > 0)
                        {
                            var next = operatorStack.Pop();
                            if (next.Equals(ParametersStartToken))
                            {
                                break;
                            }

                            outputBuffer.Push(next);
                        }

                        if (operatorStack.Count > 0 && operatorStack.Peek() is IFunction)
                        {
                            if (operatorStack.Peek() is FunctionN)
                            {
                                var arity = arityStack.Pop();
                                arity = previous != null && previous.Equals(ParametersStartToken) ? 0 : arity;
                                outputBuffer.Push(new Arity(arity));
                            }

                            outputBuffer.Push(operatorStack.Pop());
                        }

                        break;
                    default:
                        outputBuffer.Push(token);
                        break;
                }
            }
        }

        while (operatorStack.Count > 0)
        {
            outputBuffer.Push(operatorStack.Pop());
        }

        return new ShuntingYard(outputBuffer);
    }
}